### ISSUE TITLE
Support fbeta score with beta=0.5 as a built-in metric

### ISSUE THREAD

**Issue by desilinguist** [EducationalTestingService/skll#486]
It would be nice to support some variants of [`fbeta_score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.fbeta_score.html#sklearn.metrics.fbeta_score) natively in SKLL - F(0.5) at the very least because of applications in GED/GEC.

**Comment by desilinguist** [comment]
@chaomenghsuan I have been thinking about how to specify the value of `beta` parameter a bit more elegantly instead of having a whole new `beta` config option. What if  we allow users to specify something like `f0.5 score_micro` in the config file and basically _parse_ out the value of `beta` from the name of the metric itself. This would also extend to existing `f1_score_*` metrics.  And then you can assign the `make_scorer()` object to `SCORERS` in `config.py` and raise any exceptions much earlier in the process than later?

I am sorry to keep making moving the goal posts here but I think this might be worth discussing. What do you think? And what does everyone else think? @mulhod @bndgyawali @aoifecahill @jbiggsets?

@ananyaganesh I think this might interact with your issue (#491) as well so your thoughts are also welcome.

**Comment by chaomenghsuan** [comment]
now `fbeta_score_average` style metrics are dynamically added into SCORERS in `config.py`. The problem is, they cannot be pre-installed in `_CLASSIFICATION_ONLY_OBJ_FUNCS` in `learner.py`. 

One possible (but not the most elegant) solution is, in `learner.py`, when `Learner.train()` checks if `grid_objective`(str, one metric function) in `_CLASSIFICATION_ONLY_OBJ_FUNCS`, at the same time checks if it's a `fbeta_score_average` style str using regex.

for example:
```python
fbeta_pt = re.compile(r'f[0-9/.]+_score_[a-z]+')
if grid_objective not in _CLASSIFICATION_ONLY_OBJ_FUNCS and not fbeta_pt.match(grid_objective):
    raise ValueError()
```

**Comment by desilinguist** [comment]
I have been thinking about this a bit. I think a better longer term solution is to add support for custom metrics just like we have for custom learners. This means we won't need to keep adding new metrics and users can use any metric they want as long as it takes two vectors as input: the labels and the predictions. 

The machinery for solution this would work as follows:

1. We create a new mutable set called `CUSTOM_METRICS` in `skll/utils/constants.py`.

2. We add a new function called `add_custom_metric()` that takes in a path to a Python file containing a function that takes two arraylikes as arguments and returns a number. Whenever this function is called, it imports the metric function into the namespace, adds a new entry to the `SCORERS` dictionary in `skll/__init__.py`, and adds the metric name to the `CUSTOM_METRICS` set.

3. We add a new configuration field called `custom_metrics` that accepts a list of paths to Python files. 

4. In `experiments.py`, we validate this list and call `add_custom_metric()` to "register" each specified custom metrics in `SCORERS` and `CUSTOM_METRICS`.

5. We modify `get_acceptable_classificiation_metrics()`  and `get_acceptable_regression_metrics()` to return not just the built-in acceptable metrics but also any custom metrics. It's the user's responsibility to make sure that they are specifying the right custom metric for the right problem type.

This machinery should work for both API as well as config-file usage.  Thoughts? @aoifecahill @mulhod @bndgyawali @chaomenghsuan?

**Comment by aoifecahill** [comment]
This sounds like a reasonable idea to me and seems more extensible. People could also share custom metrics and if we notice that some custom metrics would be useful as built-ins, we could then consider that. 

It would be nice to also be able to pass in other optional arguments (like a second array of gold labels for example to compute metrics like PRMSE). 

I still think that having a built-in fbeta_0.5 would be useful since it's pretty standard for some tasks ;)

**Comment by desilinguist** [comment]
Oh, sorry. Yes, we can still add in `f05` (`fbeta_0_5`?, python doesn't support periods in variable names). 

The problem with optional arguments always comes down to how you would represent them in the configuration file.

**Comment by desilinguist** [comment]
I will file a new issue for custom metrics.

**Comment by mulhod** [comment]
If using an argument like @aoifecahill mentioned, an array of second human scores that parallels `y_true`, I don't think it could be used as a grid objective, for example, because it would have to be passed around like `y_true` (sliced up, etc.) for each fold, which `sklearn` does not do without some significant modification. But, yes, I think this is a great idea nonetheless.

**Comment by aoifecahill** [comment]
ah yes, that makes things more complicate then. Though it could still be useful to have it as a reported metric (and not use it as a tuning objective).
